### ISSUE TITLE
Support for plugins using dot notation to indicate namespaces

### ISSUE THREAD

**Issue by kmdouglass** [napari/npe2#226]
# Summary

Should Napari provide support for plugins whose names use namespaces in dot notation to indicate ownership?

This issue is intended to consolidate discussion on this topic.

# Discussion

PyPI and the Python packaging infrastructure support namespaces to identify ownership. For example, [gp.bootstrap](https://pypi.org/project/gp.bootstrap/) is a package owned by Gael Pasgrimaud, who has multiple PyPI packages starting with the `gp.` namespace.

It appears that Napari currently does not support plugins whose names contain dots to separate namespaces. The following Napari manifest for a fictitious plugin called `kyle-douglass.my-plugin` fails validation: 

```
name: kyle-douglass.my-plugin
contributions:
  commands:
    - id: kyle-douglass.my-plugin.frame_rate_widget
      title: open my widget
      python_name: kyle_douglass.my_plugin._widget:Frame_rate_Widget
    - id: kyle-douglass.my-plugin.Time_scroller_widget
      title: add time scroller widget
      python_name: kyle_douglass.my_plugin._widget:Time_scroller_widget
  widgets:
    - command: kyle-douglass.my-plugin.frame_rate_widget 
      display_name: Plot frame rate
    - command: kyle-douglass.my-plugin.Time_scroller_widget
      display_name: Add time scroller
```

The output of `npe2 validate` is:

```console
Invalid! 2 validation errors for PluginManifest
name
  'kyle-douglass.my-plugin' is not a valid python package name. (type=value_error)
__root__
  Commands identifiers must start with the current package name None
            the following commands where found to break this assumption:
                ['kyle-douglass.my-plugin.frame_rate_widget', 'kyle-douglass.my-plugin.Time_scroller_widget']
 (type=value_error)
```

# Misc.

- Namespaces to identify ownership on PyPI are discussed in [PEP-423](https://peps.python.org/pep-0423/).
- It's possible that Napari plugins and their corresponding PyPI packages need not be the same, but this violates some suggestions in PEP-423 for consistency in package naming.
- Namespace packages are a slightly different concept than namespaces on PyPI and, I think, are unrelated: https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages

**Comment by tlambert03** [comment]
Ahh I see, so at the moment your only consideration is just in package naming for a  single (non namespace package)?  You're not yet looking for support for true namespace packages (where individual pip installable packages can contribute to the same top level namespace)?

If so, then yes that will be as easy as simply relaxing the validator on the manifest name field.

**Comment by nclack** [comment]
> as easy as simply relaxing the validator on the manifest name field

definitely supportive of this

**Comment by tlambert03** [comment]
yeah, I just took a look at `gp.bootstrap`,  this is a pretty simply, cut & dry case.  we should definitely be allowing periods in names, in accordance with the regex offered by [PEP508](https://peps.python.org/pep-0508/#names)

```
^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$
```

basically, anything that you could pass to `importlib.metadata.distribution()` should work for `manifest.name`

**Comment by tlambert03** [comment]
> Namespace packages are a slightly different concept than namespaces on PyPI and, I think, are unrelated: https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages

also looked into this.  I think true [namespace packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/) would _also_ be fine.

```
mynamespace-subpackage-a/
    setup.cfg
    mynamespace/
        subpackage_a/
            pkg_a_manifest.yml
            __init__.py

mynamespace-subpackage-b/
    setup.cfg
    mynamespace/
        subpackage_b/
            pkg_b_manifest.yml
            __init__.py
        module_b.py
```

as long as `manifest.name` was different in `pkg_b_manifest` and `pkg_a_manifest` (and they should be: since they should match the distribution name, `mynamespace-subpackage-b`, and `mynamespace-subpackage-a` respectively)

**Comment by kmdouglass** [comment]
Thanks @tlambert03 

I agree with your analysis:

- the Napari plugin name validation logic was more restrictive than the Python packaging infrastructure
- native namespace packages should also work with the more permissive validation

Cheers!

**Comment by elio01epfl** [comment]
Hi @tlambert03 
After installing npe2 0.6.0 I tried to validate a package having in the name used the dot notation for namespaces.
It gives me the following error
```console
🅇 Invalid! 1 validation error for PluginManifest
__root__
  Commands identifiers must start with the current package name
            'leb.eda-napari' the following commands where found to break this assumption:
                ['leb.eda-napari.frame_rate_widget', 'leb.eda-napari.Time_scroller_widget']
 (type=value_error)
```
This is my napari.yaml file:
```yaml
name: leb.eda-napari
contributions:
  commands:
    - id: leb.eda-napari.frame_rate_widget
      title: open my widget
      python_name: leb.eda_napari._widget:Frame_rate_Widget
    - id: leb.eda-napari.Time_scroller_widget
      title: add time scroller widget
      python_name: leb.eda_napari._widget:Time_scroller_widget
  widgets:
    - command: leb.eda-napari.frame_rate_widget 
      display_name: Plot frame rate
    - command: leb.eda-napari.Time_scroller_widget
      display_name: Add time scroller
```

To me the package name seems to be the same to the name at the begin of the commands identifiers, have I made a mistake somewhere else or it could be a bug?

**Comment by kmdouglass** [comment]
> Hi @tlambert03 After installing npe2 0.6.0 I tried to validate a package having in the name used the dot notation for namespaces. It gives me the following error
> 
> ```
> 🅇 Invalid! 1 validation error for PluginManifest
> __root__
>   Commands identifiers must start with the current package name
>             'leb.eda-napari' the following commands where found to break this assumption:
>                 ['leb.eda-napari.frame_rate_widget', 'leb.eda-napari.Time_scroller_widget']
>  (type=value_error)
> ```
> 
> This is my napari.yaml file:
> 
> ```yaml
> name: leb.eda-napari
> contributions:
>   commands:
>     - id: leb.eda-napari.frame_rate_widget
>       title: open my widget
>       python_name: leb.eda_napari._widget:Frame_rate_Widget
>     - id: leb.eda-napari.Time_scroller_widget
>       title: add time scroller widget
>       python_name: leb.eda_napari._widget:Time_scroller_widget
>   widgets:
>     - command: leb.eda-napari.frame_rate_widget 
>       display_name: Plot frame rate
>     - command: leb.eda-napari.Time_scroller_widget
>       display_name: Add time scroller
> ```
> 
> To me the package name seems to be the same to the name at the begin of the commands identifiers, have I made a mistake somewhere else or it could be a bug?

I this is a bug because the command validation splits the commands at each dot into a list and assumes that the first element in the list is the package name: https://github.com/napari/npe2/blob/56ebc45fb9e935492c8b4e215ca048df68e4422e/npe2/manifest/schema.py#L186

**Comment by tlambert03** [comment]
Ugh, sorry.  I really should have added a test for parsing from the full manifest.  Will fix and rerelease today

**Comment by tlambert03** [comment]
ok, fixed in 0.6.1 ... but it looks like PyPI is having some network issues, so it may not be installable for a little bit

**Comment by kmdouglass** [comment]
Thanks for the quick fix @tlambert03 ! We appreciate it.
